pkg/visitors/createfv.go

```diff
@@ -0,0 +1,240 @@
+			return fmt.Errorf("Cannot create FV ending at %#x (%#x + %#x), BIOS region ends at %#x", v.AbsOffset+v.Size, v.AbsOffset, v.Size, end)
+		}
+
+		// Manually visit children, we want the index
+		idx := -1
+		var bp *uefi.BIOSPadding
+	L:
+		for i, e := range f.Elements {
+			switch f := e.Value.(type) {
+			case *uefi.BIOSPadding:
+				bp = f
+				bpOffset := offset + f.Offset
+				bpEnd := bpOffset + uint64(len(f.Buf()))
+				if v.AbsOffset >= bpOffset && v.AbsOffset+v.Size <= bpEnd {
+
+					idx = i
+					break L
+				}
+			}
+		}
+
+		if idx == -1 {
+			return fmt.Errorf("Cannot create FV at %#x (+%#x) no matching BIOS Pad found", v.AbsOffset, v.Size)
+		}
+		v.found = true
+		fv, err := createEmptyFirmwareVolume(v.AbsOffset-offset, v.Size, &v.Name)
+		if err != nil {
+			return err
+		}
+		return insertFVinBP(f, v.AbsOffset-offset, bp, idx, fv)
+	}
+
+	return f.ApplyChildren(v)
+}
+
+func insertFVinBP(br *uefi.BIOSRegion, offset uint64, bp *uefi.BIOSPadding, idx int, fv *uefi.FirmwareVolume) error {
+	// Copy the Elements before the modified BIOS Pad
+	newElements := make([]*uefi.TypedFirmware, idx, len(br.Elements))
+	copy(newElements, br.Elements[:idx])
+
+	bpBuf := bp.Buf()
+	// keep part of the BIOS Pad before the new fv if needed
+	if bp.Offset < offset {
+		hbp, err := uefi.NewBIOSPadding(bpBuf[:offset-bp.Offset], bp.Offset)
+		if err != nil {
+			return err
+		}
+		newElements = append(newElements, uefi.MakeTyped(hbp))
+	}
+
+	// Add the FV
+	newElements = append(newElements, uefi.MakeTyped(fv))
+
+	// keep part of the BIOS Pad after the new fv if needed
+	if offset-bp.Offset+fv.Length < uint64(len(bpBuf)) {
+		tbp, err := uefi.NewBIOSPadding(bpBuf[offset-bp.Offset+fv.Length:], offset+fv.Length)
+		if err != nil {
+			return err
+		}
+		newElements = append(newElements, uefi.MakeTyped(tbp))
+	}
+
+	// Keep the remaining Elements in the BIOS Region
+	if idx+1 < len(br.Elements) {
+		newElements = append(newElements, br.Elements[idx+1:]...)
+	}
+	br.Elements = newElements
+	return nil
+}
+
+func createEmptyFirmwareVolume(fvOffset, size uint64, name *guid.GUID) (*uefi.FirmwareVolume, error) {
+	// TODO: can this be refactored with the code in repack.go and assemble.go ?
+	fv := &uefi.FirmwareVolume{} // new Firmware Volume
+	// Set up volume header first.
+	fv.FileSystemGUID = *uefi.FFS2
+	fv.Signature = binary.LittleEndian.Uint32([]byte("_FVH"))
+	// TODO: retrieve all details from (all) other fv in BIOS Region
+	fv.Attributes = 0x0004FEFF
+	fv.Revision = 2
+	// Create Blocks
+	fv.Blocks = make([]uefi.Block, 2)
+	fv.Blocks[0] = uefi.Block{Size: 4096, Count: uint32(size / 4096)}
+	fv.Blocks[1] = uefi.Block{}
+	// Calculate the HeaderLen field
+	fv.HeaderLen = uint16(uefi.FirmwareVolumeFixedHeaderSize + int(unsafe.Sizeof(uefi.Block{}))*len(fv.Blocks))
+
+	fv.DataOffset = uint64(fv.HeaderLen) // unless we add the extended header
+	fv.Length = size
+
+	if name != nil {
+		// TODO: should the extended header offset be computed ?
```

Comment:
Probably, if we ever want to have weird block sizes, not that I've ever seen a system with more than 2 block entries. This is ok for now.

Reply:
updated the TODO